server/get_payload.go

```diff
@@ -174,141 +271,318 @@ func verifyPayload[P Payload](payload P, log *logrus.Entry, response *builderApi
-		})
-	case *eth2ApiV1Electra.SignedBlindedBeaconBlock:
-		return log.WithFields(logrus.Fields{
-			"ua":         userAgent,
-			"slot":       block.Message.Slot,
-			"blockHash":  block.Message.Body.ExecutionPayloadHeader.BlockHash.String(),
-			"parentHash": block.Message.Body.ExecutionPayloadHeader.ParentHash.String(),
-			"slotUID":    slotUID,
-		})
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Serialization Functions
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// canUnmarshalSSZ is an interface for types that can unmarshal SSZ
+type canUnmarshalSSZ interface {
+	UnmarshalSSZ(input []byte) error
+}
+
+// convertSSZToJSON converts SSZ-encoded bytes to JSON based on the given ethConsensusVersion
+func convertSSZToJSON(ethConsensusVersion string, sszBytes []byte) ([]byte, error) {
+	var block canUnmarshalSSZ
+	switch ethConsensusVersion {
+	case EthConsensusVersionBellatrix:
+		block = new(eth2ApiV1Bellatrix.SignedBlindedBeaconBlock)
+	case EthConsensusVersionCapella:
+		block = new(eth2ApiV1Capella.SignedBlindedBeaconBlock)
+	case EthConsensusVersionDeneb:
+		block = new(eth2ApiV1Deneb.SignedBlindedBeaconBlock)
+	case EthConsensusVersionElectra:
+		block = new(eth2ApiV1Electra.SignedBlindedBeaconBlock)
+	default:
+		return nil, errInvalidForkVersion
 	}
-	return nil
+
+	// Unmarshal the SSZ-encoded bytes into the block
+	if err := block.UnmarshalSSZ(sszBytes); err != nil {
+		return nil, err
+	}
+
+	// Re-encode the block as JSON
+	return json.Marshal(block)
+}
+
+// decodeSignedBlindedBeaconBlock will decode the request block in either JSON or SSZ.
+// Note: when decoding JSON, we must attempt decoding from newest to oldest fork version.
+func decodeSignedBlindedBeaconBlock(in []byte, contentType, ethConsensusVersion string, out *eth2Api.VersionedSignedBlindedBeaconBlock) error {
+	switch contentType {
+	case MediaTypeOctetStream:
+		if ethConsensusVersion != "" {
+			switch ethConsensusVersion {
+			case EthConsensusVersionBellatrix:
+				out.Version = spec.DataVersionBellatrix
+				out.Bellatrix = new(eth2ApiV1Bellatrix.SignedBlindedBeaconBlock)
+				return out.Bellatrix.UnmarshalSSZ(in)
+			case EthConsensusVersionCapella:
+				out.Version = spec.DataVersionCapella
+				out.Capella = new(eth2ApiV1Capella.SignedBlindedBeaconBlock)
+				return out.Capella.UnmarshalSSZ(in)
+			case EthConsensusVersionDeneb:
+				out.Version = spec.DataVersionDeneb
+				out.Deneb = new(eth2ApiV1Deneb.SignedBlindedBeaconBlock)
+				return out.Deneb.UnmarshalSSZ(in)
+			case EthConsensusVersionElectra:
+				out.Version = spec.DataVersionElectra
+				out.Electra = new(eth2ApiV1Electra.SignedBlindedBeaconBlock)
+				return out.Electra.UnmarshalSSZ(in)
+			default:
+				return errInvalidForkVersion
+			}
+		} else {
+			return types.ErrMissingEthConsensusVersion
+		}
+	case MediaTypeJSON:
```

Comment:
when decoding json, can ethConsensusVersion be used and have the decoding attempts be a fallback?
I think now all clients will start to add the consensus version as a required field in the header so this can be deprecated when there's no longer a backward compatibility issue

Reply:
This is an interesting idea, but I'm not sure it's worth it. If a client has set `Eth-Consensus-Version`, their preferred content type will almost certainly be SSZ. So I don't think this would ever be used. Also, `Eth-Consensus-Version` [is an optional field](https://github.com/ethereum/builder-specs/blob/1569fc45d43e3ed2bed94ec6788c3f8fa7d2a4b2/apis/builder/header.yaml#L52) and I'm unsure they would set it when sending JSON.

Reply:
> Eth-Consensus-Version [is an optional field](https://github.com/ethereum/builder-specs/blob/1569fc45d43e3ed2bed94ec6788c3f8fa7d2a4b2/apis/builder/header.yaml#L52) and I'm unsure they would set it when sending JSON.

to be clear on the spec side, the intend is to make this mandatory even for JSON similar to how it works on the beacon-api but as we can't expect them to set it right now, it's safer to not rely on it in case of JSON requests / responses

there are some performance improvements that even apply in case of JSON, eg. you can use a predefined schema to parse the data which is generally 2-3x more performant than a general purpose JSON parser